File: web/src/app/services/subject-assignment.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

export interface SubjectAssignment {
    serialNo: string;
    subjectId: string;
}

export interface SubjectAssignmentPage {
    items: SubjectAssignment[];
    nextCursor: string | null;
}

@Injectable({ providedIn: 'root' })
export class SubjectAssignmentService {
    private http = inject(HttpClient);

    listAssignments(apiBase: string, cursor?: string | null, size = 20,
                    serialFilter?: string | null, subjectFilter?: string | null): Observable<SubjectAssignmentPage> {
        const params: Record<string, string | number> = { size };
        if (cursor) params['cursor'] = cursor;
        if (serialFilter) params['serialFilter'] = serialFilter;
        if (subjectFilter) params['subjectFilter'] = subjectFilter;
        return this.http.get<SubjectAssignmentPage>(`${apiBase}/tenant-admin/subjects`, { params });
    }

    createAssignment(apiBase: string, serialNo: string, subjectId: string): Observable<SubjectAssignment> {
        return this.http.post<SubjectAssignment>(`${apiBase}/tenant-admin/subjects`, { serialNo, subjectId });
    }

    updateAssignment(apiBase: string, serialNo: string, newSubjectId: string): Observable<SubjectAssignment> {
        return this.http.put<SubjectAssignment>(
            `${apiBase}/tenant-admin/subjects/${encodeURIComponent(serialNo)}`,
            { serialNo, subjectId: newSubjectId }
        );
    }

    deleteAssignment(apiBase: string, serialNo: string): Observable<void> {
        return this.http.delete<void>(`${apiBase}/tenant-admin/subjects/${encodeURIComponent(serialNo)}`);
    }
}

File: backend/tenant/src/main/java/jct/pillorganizer/tenant/service/SubjectAssignmentService.java
package jct.pillorganizer.tenant.service;

import io.micronaut.core.annotation.Nullable;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import jct.pillorganizer.tenant.model.trial.SubjectAssignment;
import jct.pillorganizer.tenant.repo.SubjectAssignmentRepository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Singleton
public class SubjectAssignmentService {

    @Inject
    SubjectAssignmentRepository subjectAssignmentRepository;

    public SubjectAssignment createAssignment(String serialNo, String subjectId) {
        SubjectAssignment assignment = new SubjectAssignment();
        assignment.setId(UUID.randomUUID());
        assignment.setSerialNo(serialNo);
        assignment.setSubjectId(subjectId);
        return subjectAssignmentRepository.save(assignment);
    }

    public Optional<SubjectAssignment> getBySerialNo(String serialNo) {
        return subjectAssignmentRepository.findBySerialNo(serialNo);
    }

    public Optional<SubjectAssignment> getBySubjectId(String subjectId) {
        return subjectAssignmentRepository.findBySubjectId(subjectId);
    }

    public SubjectAssignment updateAssignment(SubjectAssignment existing, String newSubjectId) {
        existing.setSubjectId(newSubjectId);
        return subjectAssignmentRepository.update(existing);
    }

    public void deleteAssignment(SubjectAssignment existing) {
        subjectAssignmentRepository.delete(existing);
    }

    public List<SubjectAssignment> listAssignments(@Nullable String serialFilter,
                                                   @Nullable String subjectFilter,
                                                   @Nullable String cursorSerial,
                                                   int size) {
        return subjectAssignmentRepository.listAssignments(serialFilter, subjectFilter, cursorSerial, size);
    }
}
